### PR TITLE
Fix deadlock in the cacher

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #10475

Do NOT merge together with ##12721 - I need to rebase one of those before merging the other.

cc @davidopp

### CODE REVIEW

File: pkg/client/cache/watch_cache.go [hunk 1]
```diff
@@ -225,9 +225,11 @@ func (w *WatchCache) SetOnEvent(onEvent func(watch.Event)) {
 	w.onEvent = onEvent
 }
 
-func (w *WatchCache) GetAllEventsSince(resourceVersion uint64) ([]watch.Event, error) {
+// This function locks the whole cache for and doesn't release the lock on return.
+// The caller should explicitly unlock the cache by calling the returned function.
+func (w *WatchCache) GetAllEventsSinceAndLock(resourceVersion uint64) ([]watch.Event, func(), error) {
 	w.RLock()
-	defer w.RUnlock()
+	unlockFunc := w.RUnlock
```
Comment (bprashanth) [inline pkg/client/cache/watch_cache.go]: can't you just make this function thread unsafe and mandate the caller takes the rlock? I think distributed locking will be a pain to debug, especially in the watch event driven critical path of a high qps server.

Comment (wojtek-t) [inline pkg/client/cache/watch_cache.go]: Done.

**Comment by wojtek-t** [comment]
@bprashanth PTAL

**Comment by wojtek-t** [comment]
@bprashanth - friendly ping

**Comment by bprashanth** [comment]
LGTM

**Comment by saad-ali** [comment]
> Do NOT merge together with #12721 - I need to rebase one of those before merging the other.

#12721 was merged 2 days ago

**Pull request merged**